### PR TITLE
Add ENV vars to railtie

### PR DESCRIPTION
**Pull request by dudo**
We've been having some issues with load order of this railtie. We have a gem that loads several other gems (including flipper, obviously) that has its own railtie. 

The load order of these railties in railties in railties is fragile, and we feel that adding some ENV vars would alleviate this fragility.

If this is something you're ok with, I'll add documentation and tests around this change.

### CODE REVIEW

**Review by jnunemaker — COMMENTED** [review]
@bkeepers any thoughts? Most seem fine. The only part I'm in the fence about is the constantize.

File: lib/flipper/railtie.rb [hunk 1]
```diff
@@ -2,11 +2,11 @@ module Flipper
...
     config.before_configuration do
       config.flipper = ActiveSupport::OrderedOptions.new.update(
-        env_key: "flipper",
-        memoize: true,
-        preload: true,
-        instrumenter: ActiveSupport::Notifications,
-        log: true
+        env_key: ENV.fetch('FLIPPER_ENV_KEY', 'flipper'),
+        memoize: ENV.fetch('FLIPPER_MEMOIZE', 'true') == 'true',      
+        preload: ENV.fetch('FLIPPER_PRELOAD', 'true') == 'true',
```
Comment (bkeepers) [inline lib/flipper/railtie.rb:7]: I'd like to see the casting here be a little more robust. Maybe use `ActiveModel::Type::Boolean.new.cast(value)`?

**Review by bkeepers — COMMENTED** [review]
I love ENV vars and this seems like a reasonable approach to me, however, I'd also like to hear more about the issue you're having in case other people are having the same issue.

File: lib/flipper/railtie.rb (hunk 1 above)
Comment (dudo) [reply lib/flipper/railtie.rb:7]: Well... this is a railtie. Probably safe to assume ActiveModel is loaded? That being said string comparison is pretty robust, no? No need to over complicate things.

**Comment by dudo** [comment]
@jnunemaker `constantize` feels like the safest bet, since we're in a railtie. The default is loading `ActiveSupport`, so if we expect that to be ok, then we should be able to lean on its helpers?

**Comment by dudo** [comment]
Sounds like ya'll are ok with the change, though - I'll update the docs and add specs.

**Comment by dudo** [comment]
The readme doesn't feel like the right place for this config (as that mainly applies to Flipper usage, not config). Is the documentation open source? And while we're talking about it... can we add search to the docs?!

File: lib/flipper/railtie.rb (hunk 1 above)
Comment (dudo) [reply lib/flipper/railtie.rb:7]: For `preload`, we can provide an array... do we want to support that via the ENV?

Comment (dudo) [reply lib/flipper/railtie.rb]: Bumped to `casecmp` to catch `False`

**Comment by dudo** [comment]
@jnunemaker Any more suggestions to get this merged in?

**Comment by jnunemaker** [comment]
Sorry! I’ve been swamped. I’ll try to get this merged this week.
Appreciate the work!

On November 30, 2022, David Grevink ***@***.***> wrote:
> @jnunemaker <https://github.com/jnunemaker> Any more suggestions to
> get this merged in?
>

—
Reply to this email directly, view it on GitHub
<https://github.com/jnunemaker/flipper/pull/681#issuecomment-
1332416430>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAAAB25MQORG2VYJDBK2FFTWK547NANCNFSM6AAAAAAR4ZU5TM>.
You are receiving this because you were mentioned.Message ID:
***@***.***>

**Comment by jnunemaker** [comment]
> The readme doesn't feel like the right place for this config (as that mainly applies to Flipper usage, not config). Is the documentation open source? 

This is the downside of our doc move. The upside is now people realize that cloud exists which helps us grow that side of things and thus flipper. If you want to just drop a few paragraphs here i can move them into the docs.

> And while we're talking about it... can we add search to the docs?!

I just went through the process with algolia to add search. Should be out soon.

**Comment by jnunemaker** [comment]
@dudo https://www.flippercloud.io/docs has search now powered by algolia. We've noticed a few things we need to tweak but its good enough to release.

**Comment by dpep** [comment]
thanks for this, @dudo !

@jnunemaker : this relates to the work we did in https://github.com/jnunemaker/flipper/pull/586 regarding railtie load order issues.  unfortunately, it bit my company again during a flipper upgrade

**Comment by bkeepers** [comment]
@dpep can you explain what happened?

**Comment by dpep** [comment]
@bkeepers: essentially, we have an internal gem that configures Flipper for internal services.  it does so within a rails initializer, which itself gets configured when the gem is loaded.  we noticed that the services using this gem didn't have Flipper configured as expected and we suspect a load order issue.  it's entirely possible that our convoluted setup is to blame here, but after a few failed attempts to correct the issue it feels safer to switch tactics eg. @dudo's ENV variable bypass in this PR

**Pull request merged**